apps/glusterfs/operations.go

```diff
@@ -671,6 +671,137 @@ func (vdel *BlockVolumeDeleteOperation) Finalize() error {
 	})
 }
 
+// DeviceRemoveOperation is a phony-ish operation that exists
+// primarily to a) know that set state was being performed
+// and b) to serve as a starting point for a more proper
+// operation in the future.
+type DeviceRemoveOperation struct {
+	OperationManager
+	DeviceId  string
+	allocator Allocator
+}
+
+// Note: passing this allocator here a big hack, but its a temporary
+// workaround for how brick relocation currently works
+func NewDeviceRemoveOperation(
+	deviceId string, allocator Allocator, db wdb.DB) *DeviceRemoveOperation {
+
+	return &DeviceRemoveOperation{
+		OperationManager: OperationManager{
+			db: db,
+			op: NewPendingOperationEntry(NEW_ID),
+		},
+		DeviceId:  deviceId,
+		allocator: allocator,
+	}
+}
+
+func (dro *DeviceRemoveOperation) Label() string {
+	return "Remove Device"
+}
+
+func (dro *DeviceRemoveOperation) ResourceUrl() string {
+	return ""
+}
+
+func (dro *DeviceRemoveOperation) Build(allocator Allocator) error {
+	return dro.db.Update(func(tx *bolt.Tx) error {
+		d, err := NewDeviceEntryFromId(tx, dro.DeviceId)
+		if err != nil {
+			return err
+		}
+		txdb := wdb.WrapTx(tx)
+
+		// If the device has no bricks, just change the state and we are done
+		if err := d.markFailed(txdb); err == nil {
+			// device was empty and is now marked failed
+			return nil
+		} else if err != ErrConflict {
+			// we hit some sort of unexpected error
+			return err
+		}
+		// if we're here markFailed couldn't apply due to conflicts
+		// we don't need to actually record anything in the db
+		// because this is not a long running operation
+
+		if p, err := PendingOperationsOnDevice(txdb, d.Info.Id); err != nil {
+			return err
+		} else if p {
+			logger.LogError("Found operations still pending on device."+
+				" Can not remove device %v at this time.",
+				d.Info.Id)
+			return ErrConflict
+		}
+
+		dro.op.RecordRemoveDevice(d)
+		if e := dro.op.Save(tx); e != nil {
+			return e
+		}
+		return nil
+	})
+}
+
+func (dro *DeviceRemoveOperation) deviceId() (string, error) {
+	if len(dro.op.Actions) == 0 {
+		// we intentionally avoid recording any actions when all needed bits
+		// were taken care of in Build. There's nothing more to do here.
+		return "", nil
+	}
+	if dro.op.Actions[0].Change != OpRemoveDevice {
+		return "", fmt.Errorf("Unexpected action (%v) on DeviceRemoveOperation pending op",
+			dro.op.Actions[0].Change)
+	}
+	return dro.op.Actions[0].Id, nil
+}
+
+func (dro *DeviceRemoveOperation) Exec(executor executors.Executor) error {
+	id, err := dro.deviceId()
+	if err != nil {
+		return err
+	}
+	if id == "" {
+		return nil
+	}
+
+	var d *DeviceEntry
+	if e := dro.db.Update(func(tx *bolt.Tx) error {
```

Comment:
Why `Update()` instead of `View()` here?

Reply:
Good catch, There's no reason this should be an Update.